tests/restore_state.yml

```diff
@@ -11,3 +11,18 @@
 
 - name: Restore services
   import_tasks: restore_services_state.yml
+
+- name: restart NetworkManager
+  tags:
+    - 'tests::cleanup'
+  service:
+    name: NetworkManager
```

Comment:
Restarting NetworkManager is not the right thing to reload profiles from disk. In fact, the only reason to ever restart NetworkManager service is when upgrading the package (without doing a reboot, which is already dirty), or when there is a serious bug and NM is dead (in which case, it is merely a workaround but not the solution).

To reload profiles either
```
    busctl call org.freedesktop.NetworkManager \
            /org/freedesktop/NetworkManager/Settings \
            org.freedesktop.NetworkManager.Settings \
            ReloadConnections 
```
or
```
    nmcli connection reload
```

Reply:
The restart is a workaround to make NetworkManager forget that devices might be unmanaged (when earlier initscripts profiles with NM_CONTROLLED=no were deployed). Should your commands be enough in this case?

Reply:
Thanks for your comment. To give you a context, the problem is, we want to run the tests against localhost, in a CI setting where we want fast results and we can't spawn VMs. The problem is now interaction between the various tests. We implemented cleanups, but we found out that there is still a problem if a test playbook for initscripts runs, creates a connection (veth or bridge) and removes it. As NetworkManager is continuously running during that, it sees the connection and remembers that it should ignore it, and it persists in its state even after the connection is removed. This leads to subsequent tests for Networkmanager failing if they create a connection of the same name - NetworkManager refuses to create it. @tyll suggested restart as a solution to this (and it works). @tyll, will the approach proposed by @thom311 work as well?

Reply:
Not sure I understand, but `nmcli connection reload` is supposed to correctly handle changes to `NM_CONTROLLED=no`.

If you write `NM_CONTROLLED=no` to an ifcfg file and issue `ifup eth0`, then this implicitly does a `nmcli connection load /etc/sysconfig/network-scripts/ifcfg-eth0`, which results in the device becoming unmanaged. If you remove the `NM_CONTROLLED=no` line again and issue `nmcli connection reload` (or load), then the reason why the device was unmanaged is gone and it should become managed again.


> This leads to subsequent tests for Networkmanager failing if they create a connection of the same name - NetworkManager refuses to create it.

I think, NM wouldn't refuse to create the profile, it would fail to activate a profile on the device (if that device is unmanaged due to `NM_CONTROLLED=no`). But yeah... the effect is the same. It would fail.

Reply:
We don't remove the line, we remove the whole profile.

Reply:
(the ifcfg script)

Reply:
> We don't remove the line, we remove the whole profile.

Sure. `nmcli connection reload` is supposed to cover that as well.

Reply:
> Not sure I understand, but `nmcli connection reload` is supposed to correctly handle changes to `NM_CONTROLLED=no`.
> 
> If you write `NM_CONTROLLED=no` to an ifcfg file and issue `ifup eth0`, then this implicitly does a `nmcli connection load /etc/sysconfig/network-scripts/ifcfg-eth0`, which results in the device becoming unmanaged. If you remove the `NM_CONTROLLED=no` line again and issue `nmcli connection reload` (or load), then the reason why the device was unmanaged is gone and it should become managed again.

Is there a way to unload a profile when we remove the profile without doing a generic reload? Then we could do this for persistent_state:absent with initscripts to ensure the state is also wiped from NM. Does a load for the path trigger it when the file does not exist anymore?